src/Cli/dotnet/Commands/Tool/Execute/ToolExecuteCommand.cs

```diff
@@ -25,10 +34,60 @@ internal class ToolExecuteCommand(ParseResult result) : CommandBase(result)
         private readonly bool _interactive = result.GetValue(ToolExecuteCommandParser.InteractiveOption);
         private readonly VerbosityOptions _verbosity = result.GetValue(ToolExecuteCommandParser.VerbosityOption);
         private readonly bool _yes = result.GetValue(ToolExecuteCommandParser.YesOption);
+
+        //  TODO: Does result.OptionValuesToBeForwarded work here?
+        private readonly IToolPackageDownloader _toolPackageDownloader = ToolPackageFactory.CreateToolPackageStoresAndDownloader(
+                additionalRestoreArguments: result.OptionValuesToBeForwarded(ToolExecuteCommandParser.GetCommand())).Item3;
+
+        //  TODO: Make sure to add these options to the command
+        private readonly RestoreActionConfig _restoreActionConfig = new RestoreActionConfig(DisableParallel: result.GetValue(ToolCommandRestorePassThroughOptions.DisableParallelOption),
+            NoCache: result.GetValue(ToolCommandRestorePassThroughOptions.NoCacheOption) || result.GetValue(ToolCommandRestorePassThroughOptions.NoHttpCacheOption),
+            IgnoreFailedSources: result.GetValue(ToolCommandRestorePassThroughOptions.IgnoreFailedSourcesOption),
+            Interactive: result.GetValue(ToolCommandRestorePassThroughOptions.InteractiveRestoreOption));
+
+        //  TODO: Use prerelease
         private readonly bool _prerelease = result.GetValue(ToolExecuteCommandParser.PrereleaseOption);
 
+        private readonly ToolManifestFinder _toolManifestFinder = toolManifestFinder ?? new ToolManifestFinder(new DirectoryPath(currentWorkingDirectory ?? Directory.GetCurrentDirectory()));
+
         public override int Execute()
         {
+            VersionRange versionRange = _parseResult.GetVersionRange();
+            PackageId packageId = new PackageId(_packageToolIdentityArgument.Id);
+
+            //  Look in local tools manifest first, but only if version is not specified
+            if (versionRange == null)
+            {
+                var localToolsResolverCache = new LocalToolsResolverCache();
+
+                if (_toolManifestFinder.TryFindPackageId(packageId, out var toolManifestPackage))
+                {
+                    var toolPackageRestorer = new ToolPackageRestorer(
+                        _toolPackageDownloader,
+                        _sources,
+                        overrideSources: [],
+                        _verbosity,
+                        _restoreActionConfig,
+                        localToolsResolverCache,
+                        new FileSystemWrapper());
+
+                    var restoreResult = toolPackageRestorer.InstallPackage(toolManifestPackage, _configFile == null ? null : new FilePath(_configFile));
+
+                    if (!restoreResult.IsSuccess)
+                    {
```

Comment:
Should we fall back to global? Otherwise, if you expect to find it in the global tools cache, this would start failing.

Reply:
No, I don't think we want to mix in global tools with this.  Local tools and one-shot execution both use the NuGet global packages folder, so if the local tool restore failed it's not likely that downloading the tool for one-shot execution would succeed.